File: java/matrix_chain.java
import java.util.*;

class matrix_chain{

	public static void mcs(int p[],int n,int m[][],int s[][]){
		int i,j,l,q,k;
		for(i=1;i<n;i++){
			m[i][i]=0;
			s[i][i]=1;
		}
		for(l=2;l<n;l++){
			for(i=1;i<n-l+1;i++){
				j=i+l-1;

				m[i][j]=Integer.MAX_VALUE;
				for(k=i;k<=j-1;k++){
					q=m[i][k]+m[k+1][j]+(p[i-1]*p[k]*p[j]);
					if(q<m[i][j]){
						m[i][j]=q;
						s[i][j]=k;
					}
				}
			}
		}
	}


	public static void print(int s[][],int i,int j){

		if(i==j){
			System.out.print("A"+i);
		}
		else{
			System.out.print("(");
			print(s,i,s[i][j]);
			print(s,s[i][j]+1,j);
			System.out.print(")");
		}
	}

	public static void main(String args[]){
		Scanner sc=new Scanner(System.in);

		System.out.println("Enter the number of elements:\n");
		
		int n=sc.nextInt();
		System.out.println("Enter "+n+" elements:\n");
		int p[]=new int[n];
		int m[][]=new int[n+1][n+1];
		int s[][]=new int[n+1][n+1];
		
		for(int i=0;i<n;i++){
			p[i]=sc.nextInt();
		}
		int i,j;
		
		mcs(p,n,m,s);

		for(i=1;i<n;i++){
			for(j=1;j<n;j++)
				System.out.print(""+m[i][j]+"\t");
			System.out.print("\n");
		}

		System.out.println("\n\n");
		for(i=1;i<n;i++){
			for(j=1;j<n;j++)
				System.out.print(""+s[i][j]+"\t");
			System.out.print("\n");
		}

		print(s,1,n-1);
	}

}

File: python/matrix_chain.py
import sys

def mcs(p,n,m,s):
	for i in range(1,n):
		m[i][i]=0
		s[i][i]=1
	
	for l in range(2,n):
		for i in range(1,n-l+1):
			j=i+l-1

			m[i][j]=sys.maxsize
			for k in range(i,j):
				q=m[i][k]+m[k+1][j]+(p[i-1]*p[k]*p[j])
				if(q<m[i][j]):
					m[i][j]=q
					s[i][j]=k
				
			
def prin(s,i,j):
	if(i==j):
		print("A"+str(i),end="")
		
	else:
		print("(",end="")
		prin(s,i,s[i][j])
		prin(s,s[i][j]+1,j)
		print(")",end="")
				
	

print("Enter the value of n:")
n=int(input())
li=[]

print("Enter "+str(n)+" elements:")
for i in range(n):
	li.append(int(input()))

m=[]
s=[]
for i in range(n):
	q=[]
	p=[]
	for j in range(n):
		q.append(0)
		p.append(0)

	m.append(q)
	s.append(p)

mcs(li,n,m,s)
for i in range(1,n):
	for j in range(1,n):
		print(str(m[i][j])+"\t",end="")
	print("\n",end="")

print("\n\n")
for i in range(1,n):
	for j in range(1,n):
		print(str(s[i][j])+"\t",end="")
	print("\n",end="")
prin(s,1,n-1)